diff -d -u -r dvb/si/datetime.h dvb/si/datetime.h
--- dvb/si/datetime.h	2012-01-08 20:11:32.000000000 +0100
+++ dvb/si/datetime.h	2014-06-14 15:00:42.643172561 +0200
@@ -31,6 +31,12 @@
 #include <time.h>   /* gmtime_r, time_t */
 #include <stdio.h>  /* sprintf */
 
+// Android: include 64-bit time_t, which offers a timegm function!
+// modified from http://trac.rtmpd.com/browser/trunk/sources/common/src/platform/android/timegm.cpp
+#ifdef __ANDROID__
+#include <time64.h>
+#endif
+
 /*
  * Normative references:
  *  - ETSI EN 300 468 V1.11.1 (2010-04) (SI in DVB systems)
@@ -88,7 +94,29 @@
 
     dvb_time_decode_bcd(bcd, NULL, &tm.tm_hour, &tm.tm_min, &tm.tm_sec);
 
+#ifdef _WIN32
+	// _mkgmtime would be Windows equivalent, but is not supported by MinGW!
+    // Therefore, the global variables described here are used: http://msdn.microsoft.com/en-US/library/90s5c885.aspx
+
+	// backup current setting + set UTC
+	long tz = _timezone;
+	int dl = _daylight;
+	_timezone = 0;
+	_daylight = 0;
+
+	time_t result = mktime(&tm);
+
+	// restore settings
+	_timezone = tz;
+	_daylight = dl;
+
+	return result;
+#elif defined(__ANDROID__)
+	// using 64-bit timegm, but returning 32-bit time_t!
+	return timegm64(&tm);
+#else
     return timegm(&tm);
+#endif
 }
 
 #define dec2bcd(__dec) (int)((((__dec)/10) << 4) + (__dec) % 10)
@@ -97,7 +125,12 @@
     uint16_t mjd = 0;
     uint32_t bcd = 0;
 
+#ifdef _WIN32
+    // Windows: no gmtime_r, but gmtime also thread-safe (see: http://msdn.microsoft.com/en-us/library/0z9czt0w.aspx)
+    tm = *(gmtime(&ts));
+#else
     gmtime_r(&ts, &tm);
+#endif
 
     mjd = dvb_mjd_set(tm.tm_year, tm.tm_mon + 1, tm.tm_mday);
 
@@ -129,7 +162,14 @@
     if (tm == NULL)
         tm = &tm_local;
     time_t ts = dvb_time_decode_UTC(UTC_time);
+
+#ifdef _WIN32
+    // Windows: no gmtime_r, but gmtime also thread-safe (see: http://msdn.microsoft.com/en-us/library/0z9czt0w.aspx)
+    tm = gmtime(&ts);
+#else
     tm = gmtime_r(&ts, tm);
+#endif
+
     if (output)
         sprintf(output, "%04d-%02d-%02d %02d:%02d:%02d UTC",
             tm->tm_year+1900, tm->tm_mon+1, tm->tm_mday,
diff -d -u -r dvb/si/desc_44.h dvb/si/desc_44.h
--- dvb/si/desc_44.h	2012-01-08 20:11:32.000000000 +0100
+++ dvb/si/desc_44.h	2014-06-14 11:01:47.943624541 +0200
@@ -128,14 +128,14 @@
     switch (i_print_type) {
     case PRINT_XML:
         pf_print(opaque,
-             "<CABLE_DESC frequency=\"%"PRIu64"\" fecouter=\"%s\" modulation=\"%s\" symbolrate=\"%u\" fecinner=\"%s\"/>",
+             "<CABLE_DESC frequency=\"%" PRIu64 "\" fecouter=\"%s\" modulation=\"%s\" symbolrate=\"%u\" fecinner=\"%s\"/>",
              desc44_get_frequency(p_desc), psz_fecouter, psz_modulation,
              desc44_get_symbolrate(p_desc),
              dvb_delivery_get_fec(desc43_get_fecinner(p_desc)));
         break;
     default:
         pf_print(opaque,
-             "    - desc 44 dvb-c frequency=%"PRIu64" Hz fecouter=%s modulation=%s symbolrate=%u fecinner=%s",
+             "    - desc 44 dvb-c frequency=%" PRIu64 " Hz fecouter=%s modulation=%s symbolrate=%u fecinner=%s",
              desc44_get_frequency(p_desc), psz_fecouter, psz_modulation,
              desc44_get_symbolrate(p_desc),
              dvb_delivery_get_fec(desc43_get_fecinner(p_desc)));
diff -d -u -r dvb/si/desc_45.h dvb/si/desc_45.h
--- dvb/si/desc_45.h	2012-01-08 20:11:32.000000000 +0100
+++ dvb/si/desc_45.h	2014-06-14 11:00:14.779627479 +0200
@@ -74,14 +74,14 @@
 
 static inline char *desc45_get_service_id_txt(uint8_t i_service_id)
 {
-    return i_service_id == 0x00 ? "reserved" :
+    return (char*) (i_service_id == 0x00 ? "reserved" :
            i_service_id == 0x01 ? "EBU teletext" :
            i_service_id == 0x02 ? "inverted teletext" :
            i_service_id == 0x03 ? "reserved" :
            i_service_id == 0x04 ? "VPS" :
            i_service_id == 0x05 ? "WSS" :
            i_service_id == 0x06 ? "Closed Captioning" :
-           i_service_id == 0x07 ? "Mono 4:2:2 samples" : "reserved";
+           i_service_id == 0x07 ? "Mono 4:2:2 samples" : "reserved");
 }
 
 static inline uint8_t desc45n_get_byte(uint8_t *p_desc_n, uint8_t idx)
diff -d -u -r dvb/si/desc_46.h dvb/si/desc_46.h
--- dvb/si/desc_46.h	2012-01-08 20:11:32.000000000 +0100
+++ dvb/si/desc_46.h	2014-06-14 11:00:14.779627479 +0200
@@ -79,12 +79,12 @@
 
 static inline char *desc46_get_teletexttype_txt(uint8_t i_type)
 {
-    return i_type == 0x00 ? "Reserved" :
+    return (char*) (i_type == 0x00 ? "Reserved" :
            i_type == 0x01 ? "Initial teletext page" :
            i_type == 0x02 ? "Teletext subtitle page" :
            i_type == 0x03 ? "Additional information page" :
            i_type == 0x04 ? "Programme schedule page" :
-           i_type == 0x05 ? "Teletext subtitle page for hearing impaired people schedule page" : "Reserved";
+           i_type == 0x05 ? "Teletext subtitle page for hearing impaired people schedule page" : "Reserved");
 }
 
 static inline void desc46n_set_teletextmagazine(uint8_t *p_desc_n,
diff -d -u -r dvb/si/desc_51.h dvb/si/desc_51.h
--- dvb/si/desc_51.h	2012-01-08 20:11:32.000000000 +0100
+++ dvb/si/desc_51.h	2014-06-14 11:00:14.783627479 +0200
@@ -107,10 +107,10 @@
 
 static inline char *desc51_get_logical_cell_presentation_info_txt(uint8_t i_lcp_info)
 {
-    return i_lcp_info == 0x00 ? "undefined" :
+    return (char*) (i_lcp_info == 0x00 ? "undefined" :
            i_lcp_info == 0x01 ? "video" :
            i_lcp_info == 0x02 ? "still picture" :
-           i_lcp_info == 0x03 ? "graphics/text" : "reserved";
+           i_lcp_info == 0x03 ? "graphics/text" : "reserved");
 }
 
 static inline uint8_t desc51n_get_elementary_cell_field_length(const uint8_t *p_desc_n)
@@ -145,11 +145,11 @@
 
 static inline char *desc51_get_cell_linkage_info_txt(uint8_t i_icl_info)
 {
-    return i_icl_info == 0x00 ? "undefined" :
+    return (char*) (i_icl_info == 0x00 ? "undefined" :
            i_icl_info == 0x01 ? "bouquet related" :
            i_icl_info == 0x02 ? "service related" :
            i_icl_info == 0x03 ? "other mosaic related" :
-           i_icl_info == 0x04 ? "event related" : "reserved";
+           i_icl_info == 0x04 ? "event related" : "reserved");
 }
 
 #define init_link_n(__const) \
diff -d -u -r dvb/si/desc_5a.h dvb/si/desc_5a.h
--- dvb/si/desc_5a.h	2012-01-08 20:11:32.000000000 +0100
+++ dvb/si/desc_5a.h	2014-06-14 11:02:15.819623662 +0200
@@ -247,7 +247,7 @@
     switch (i_print_type) {
     case PRINT_XML:
         pf_print(opaque,
-             "<TERRESTRIAL_DESC frequency=\"%"PRIu64"\" bandwidth=\"%u\" priority=\"%s\" timeslicing=\"%d\" mpefec=\"%d\" constellation=\"%s\" hierarchy=\"%s\" coderatehp=\"%s\" coderatelp=\"%s\" guard=\"%s\" transmission=\"%s\" otherfrequency=\"%d\"/>",
+             "<TERRESTRIAL_DESC frequency=\"%" PRIu64 "\" bandwidth=\"%u\" priority=\"%s\" timeslicing=\"%d\" mpefec=\"%d\" constellation=\"%s\" hierarchy=\"%s\" coderatehp=\"%s\" coderatelp=\"%s\" guard=\"%s\" transmission=\"%s\" otherfrequency=\"%d\"/>",
              desc5a_get_frequency(p_desc), i_bandwidth,
              desc5a_get_priority(p_desc) ? "HP" : "LP",
              desc5a_get_timeslicing(p_desc) ? 1 : 0,
@@ -260,7 +260,7 @@
         break;
     default:
         pf_print(opaque,
-             "    - desc 5a dvb-t frequency=%"PRIu64" Hz bandwidth=%u MHz priority=%s timeslicing=%d mpefec=%d constellation=%s hierarchy=%s coderatehp=%s coderatelp=%s guard=%s transmission=%s otherfrequency=%d",
+             "    - desc 5a dvb-t frequency=%" PRIu64 " Hz bandwidth=%u MHz priority=%s timeslicing=%d mpefec=%d constellation=%s hierarchy=%s coderatehp=%s coderatelp=%s guard=%s transmission=%s otherfrequency=%d",
              desc5a_get_frequency(p_desc), i_bandwidth,
              desc5a_get_priority(p_desc) ? "HP" : "LP",
              desc5a_get_timeslicing(p_desc) ? 1 : 0,
diff -d -u -r dvb/si/strings.h dvb/si/strings.h
--- dvb/si/strings.h	2012-01-08 20:11:32.000000000 +0100
+++ dvb/si/strings.h	2014-06-14 11:00:14.783627479 +0200
@@ -109,7 +109,7 @@
 
     for (i = 0; ppsz_dvb_encodings[i] != NULL; i++) {
         if (!strcasecmp(psz_encoding, ppsz_dvb_encodings[i])) {
-            uint8_t *p_encoded = malloc(i_length + 1);
+            uint8_t *p_encoded = (uint8_t*) malloc(i_length + 1);
             *pi_out_length = i_length + 1;
             p_encoded[0] = i;
             memcpy(p_encoded + 1, p_string, i_length);
@@ -119,7 +119,7 @@
 
     for (i = 0; ppsz_dvb_encodings10[i] != NULL; i++) {
         if (!strcasecmp(psz_encoding, ppsz_dvb_encodings10[i])) {
-            uint8_t *p_encoded = malloc(i_length + 3);
+            uint8_t *p_encoded = (uint8_t*) malloc(i_length + 3);
             *pi_out_length = i_length + 3;
             p_encoded[0] = 0x10;
             p_encoded[1] = 0x0;
@@ -142,7 +142,7 @@
                                                            &i_length);
         if (psz_encoding == NULL || !i_length) {
             /* try one-byte charset */
-            char *psz_string = malloc(i_length + 1);
+            char *psz_string = (char*) malloc(i_length + 1);
             memcpy(psz_string, p_string, i_length);
             psz_string[i_length] = '\0';
             return psz_string;
@@ -180,7 +180,7 @@
         psz1++;
     }
 
-    psz2 = psz_output = malloc(i_output_size + 1);
+    psz2 = psz_output = (char*) malloc(i_output_size + 1);
     psz1 = psz_input;
     while (*psz1) {
         switch (*psz1) {
diff -d -u -r mpeg/psi/desc_04.h mpeg/psi/desc_04.h
--- mpeg/psi/desc_04.h	2012-01-08 20:11:32.000000000 +0100
+++ mpeg/psi/desc_04.h	2014-06-14 11:00:14.783627479 +0200
@@ -94,7 +94,7 @@
 
 static inline char *desc04_get_type_txt(uint8_t i_type)
 {
-    return i_type == 0 ? "reserved" :
+    return (char*) (i_type == 0 ? "reserved" :
            i_type == 1 ? "Spatial Scalability" :
            i_type == 2 ? "SNR Scalability" :
            i_type == 3 ? "Temporal Scalability" :
@@ -103,7 +103,7 @@
            i_type == 6 ? "Private Stream" :
            i_type == 7 ? "Multi-view Profile" :
            i_type >= 8 && i_type <= 14 ? "reserved" :
-           i_type == 15 ? "base layer" : "unknown";
+           i_type == 15 ? "base layer" : "unknown");
 }
 
 static inline bool desc04_validate(const uint8_t *p_desc)
diff -d -u -r mpeg/psi/desc_24.h mpeg/psi/desc_24.h
--- mpeg/psi/desc_24.h	2012-01-08 20:11:32.000000000 +0100
+++ mpeg/psi/desc_24.h	2014-06-14 11:03:02.123622202 +0200
@@ -310,8 +310,8 @@
                  " content_reference_id_record_length=\"%u\""
                  " content_reference_id_record=\"%s\""
                  " content_time_base_indicator=\"%u\""
-                 " content_time_base_value=\"%"PRIu64"\""
-                 " metadata_time_base_value=\"%"PRIu64"\""
+                 " content_time_base_value=\"%" PRIu64 "\""
+                 " metadata_time_base_value=\"%" PRIu64 "\""
                  " content_id=\"%u\""
                  " time_base_association_data_length=\"%u\""
                  " time_base_association_data=\"%s\"/>",
@@ -336,8 +336,8 @@
                  " content_reference_id_record_length=%u"
                  " content_reference_id_record=\"%s\""
                  " content_time_base_indicator=%u"
-                 " content_time_base_value=\"%"PRIu64"\""
-                 " metadata_time_base_value=\"%"PRIu64"\""
+                 " content_time_base_value=\"%" PRIu64 "\""
+                 " metadata_time_base_value=\"%" PRIu64 "\""
                  " content_id=%u"
                  " time_base_association_data_length=%u"
                  " time_base_association_data=\"%s\"",
diff -d -u -r mpeg/psi/desc_26.h mpeg/psi/desc_26.h
--- mpeg/psi/desc_26.h	2012-01-08 20:11:32.000000000 +0100
+++ mpeg/psi/desc_26.h	2014-06-14 11:08:39.047611579 +0200
@@ -235,7 +235,7 @@
     uint8_t i, i_service_identification_record_length;
     const uint8_t *p_service_identification_record = desc26_get_service_identification_record(p_desc, &i_service_identification_record_length);
     char psz_service_identification_record[2 * 256 + 1];
-    char *extra_str = "decoder_config";
+    char *extra_str = (char*) "decoder_config";
     const uint8_t *p_extra = NULL;
     uint8_t i_extra_length = 0;
     char psz_extra[2 * 256 + 1];
@@ -248,7 +248,7 @@
         p_extra = desc26_get_decoder_config(p_desc, &i_extra_length);
     }
     if (desc26_get_decoder_config_flags(p_desc) == 3) {
-        extra_str = "dec_config_identification_record";
+        extra_str = (char*) "dec_config_identification_record";
         p_extra = desc26_get_dec_config_identification_record(p_desc, &i_extra_length);
     }
     for (i = 0; i < i_extra_length; i++)
diff -d -u -r mpeg/psi/psi.h mpeg/psi/psi.h
--- mpeg/psi/psi.h	2012-01-08 20:11:32.000000000 +0100
+++ mpeg/psi/psi.h	2014-06-14 11:00:14.783627479 +0200
@@ -151,12 +151,12 @@
 
 static inline uint8_t *psi_allocate(void)
 {
-    return malloc((PSI_MAX_SIZE + PSI_HEADER_SIZE) * sizeof(uint8_t));
+    return (uint8_t*) malloc((PSI_MAX_SIZE + PSI_HEADER_SIZE) * sizeof(uint8_t));
 }
 
 static inline uint8_t *psi_private_allocate(void)
 {
-    return malloc((PSI_PRIVATE_MAX_SIZE + PSI_HEADER_SIZE) * sizeof(uint8_t));
+    return (uint8_t*) malloc((PSI_PRIVATE_MAX_SIZE + PSI_HEADER_SIZE) * sizeof(uint8_t));
 }
 
 static inline void psi_set_tableid(uint8_t *p_section, uint8_t i_table_id)
@@ -438,7 +438,7 @@
 
 static inline uint8_t **psi_table_allocate(void)
 {
-    return malloc(PSI_TABLE_MAX_SECTIONS * sizeof(uint8_t *));
+    return (uint8_t**) malloc(PSI_TABLE_MAX_SECTIONS * sizeof(uint8_t *));
 }
 
 static inline void psi_table_init(uint8_t **pp_sections)
diff -d -u -r mpeg/ts.h mpeg/ts.h
--- mpeg/ts.h	2012-01-08 20:11:32.000000000 +0100
+++ mpeg/ts.h	2014-06-14 11:00:14.783627479 +0200
@@ -56,7 +56,7 @@
 
 static inline uint8_t *ts_allocate(void)
 {
-    return malloc(TS_SIZE * sizeof(uint8_t));
+    return (uint8_t*) malloc(TS_SIZE * sizeof(uint8_t));
 }
 
 static inline void ts_init(uint8_t *p_ts)
